ampd-handlers/src/bin/evm/handler.rs

```diff
@@ -0,0 +1,456 @@
+                .await
+                .change_context(Error::Finalizer)?;
+
+        Ok(join_all(
+            tx_hashes
+                .into_iter()
+                .map(|tx_hash| self.rpc_client.transaction_receipt(tx_hash)),
+        )
+        .await
+        .into_iter()
+        .filter_map(std::result::Result::unwrap_or_default)
+        .filter_map(|tx_receipt| {
+            if tx_receipt
+                .block_number
+                .unwrap_or(U64::MAX)
+                .le(&latest_finalized_block_height)
+            {
+                Some((tx_receipt.transaction_hash, tx_receipt))
+            } else {
+                None
+            }
+        })
+        .collect())
+    }
+
+    fn vote_msg(&self, poll_id: PollId, votes: Vec<Vote>) -> MsgExecuteContract {
+        MsgExecuteContract {
+            sender: self.verifier.clone(),
+            contract: self.voting_verifier_contract.clone(),
+            msg: serde_json::to_vec(&ExecuteMsg::Vote { poll_id, votes })
+                .expect("vote msg should serialize"),
+            funds: vec![],
+        }
+    }
+}
+
+#[async_trait]
+impl<C> EventHandler for Handler<C>
+where
+    C: EthereumClient + Send + Sync,
+{
+    type Err = Error;
+    type Event = PollStartedEvent;
+
+    async fn handle<HC: EventHandlerClient + Send + 'static>(
+        &self,
+        event: PollStartedEvent,
+        client: &mut HC,
+    ) -> Result<Vec<Any>> {
+        let PollStartedEvent {
+            poll_id,
+            source_chain,
+            source_gateway_address,
+            messages,
+            expires_at,
+            confirmation_height,
+            participants,
+        } = event;
+
+        if source_chain != self.chain {
+            return Ok(vec![]);
+        }
+
+        if !participants.contains(&self.verifier) {
+            return Ok(vec![]);
+        }
+
+        let latest_block_height = client
+            .latest_block_height()
+            .await
+            .change_context(Error::EventHandling)?;
+        if latest_block_height >= expires_at {
+            info!(poll_id = poll_id.to_string(), "skipping expired poll");
+            return Ok(vec![]);
+        }
+
+        let tx_hashes: HashSet<Hash> = messages
+            .iter()
+            .map(|msg| msg.message_id.tx_hash.into())
+            .collect();
+
+        let finalized_tx_receipts = self
+            .finalized_tx_receipts(tx_hashes, confirmation_height)
+            .await?;
+
+        let poll_id_str: String = poll_id.into();
+        let source_chain_str: String = source_chain.into();
+
+        let votes = info_span!(
+            "verify messages from an EVM chain",
+            poll_id = poll_id_str,
+            source_chain = source_chain_str,
+            message_ids = messages
+                .iter()
+                .map(|msg| msg.message_id.to_string())
+                .collect::<Vec<String>>()
+                .as_value(),
+        )
+        .in_scope(|| {
```

Comment:
nice!

Reply:
I just copied it from the old handler, I can't get the credit lol